### PR TITLE
🔏 DB-level constraints for URNs

### PR DESCRIPTION
**Pull request by rowanseymour**
* Disallow empty scheme or path 
 * Disallow identity mismatch with scheme and path

### CODE REVIEW

File: temba/contacts/migrations/0130_auto_20210105_2247.py [hunk 1]
```diff
@@ -0,0 +1,35 @@
+# Generated by Django 2.2.10 on 2021-01-05 22:47
+
+import django.db.models.expressions
+import django.db.models.functions.text
+from django.db import migrations, models
+
+
+class Migration(migrations.Migration):
```
Comment (rowanseymour) [inline temba/contacts/migrations/0130_auto_20210105_2247.py:8]:
```
BEGIN;
--
-- Create constraint non_empty_scheme_and_path on model contacturn
--
ALTER TABLE "contacts_contacturn" ADD CONSTRAINT "non_empty_scheme_and_path" CHECK (NOT (("scheme" = '' OR "path" = '')));
--
-- Create constraint identity_matches_scheme_and_path on model contacturn
--
ALTER TABLE "contacts_contacturn" ADD CONSTRAINT "identity_matches_scheme_and_path" CHECK ("identity" = (CONCAT("contacts_contacturn"."scheme", CONCAT(':', "contacts_contacturn"."path"))));
COMMIT;
```

File: temba/contacts/tests.py [hunk 2]
```diff
@@ -3595,6 +3596,18 @@ def test_get_display(self):
...
 
+    def test_empty_scheme_disallowed(self):
+        with self.assertRaises(IntegrityError):
+            ContactURN.objects.create(org=self.org, scheme="", path="1234", identity=":1234")
+
+    def test_empty_path_disallowed(self):
+        with self.assertRaises(IntegrityError):
+            ContactURN.objects.create(org=self.org, scheme="ext", path="", identity="ext:")
+
+    def test_identity_mismatch_disallowed(self):
```
Comment (rowanseymour) [inline temba/contacts/tests.py:3607]: you can only test one `IntegrityError` per unit test since the transaction is no longer usable after the first

**Comment by rowanseymour** [comment]
@nicpottier started adding a trigger/function to handle creates and updates to URNs and then talked myself out of it. The Postgres way to do this would be a generated column, i.e.

```
identity VARCHAR(255) GENERATED ALWAYS AS CONCAT(scheme, ':', path) STORED
```

But that only became possible in PG12 and support in Django is still at the discussion stage (https://code.djangoproject.com/ticket/31565, https://code.djangoproject.com/ticket/31300). No strong feelings either way but don't love the added complexity of a trigger and it means the in-memory version of the URN might not be correct.

File: temba/contacts/models.py [hunk 3]
```diff
@@ -1431,6 +1432,12 @@ def __str__(self):  # pragma: no cover
     class Meta:
         unique_together = ("identity", "org")
         ordering = ("-priority", "id")
+        constraints = [
+            models.CheckConstraint(check=~(Q(scheme="") | Q(path="")), name="non_empty_scheme_and_path"),
```
Comment (rowanseymour) [inline temba/contacts/models.py:1436]: the win with these isn't writing SQL in a weird DSL but that Django manages these thru its migrations

Comment (nicpottier) [reply temba/contacts/models.py:1436]: That's pretty crazy that their DSL has enough to support the identity check too. I feel like it is slightly odd to have another way of representing this vs SQL but sure, let's try it!

**Review by nicpottier — APPROVED** [review]

**Pull request merged**